### PR TITLE
Fix issue with labels not being saved in override lang file

### PR DESCRIPTION
**Pull request by JackBuchanan**
## Description
Make label changes in studio to use:  _override_ so that the changes are saved to ext last

Save label changes in override file in new function write_override_label_to_file
After label is saved use moduleinstaller->merge_files to that lang file are generated in ext to changes take effect immediately 

## Motivation and Context

Studio field and label changes should be saved in _override_ so that studio changes are genereated in Ext last

## How To Test This

Make studio label changes and check they are saved to: custom/Extension/modules/{module}/Ext/Language/_override_en_us.lang.php
and then generated in: custom/modules/{module}/Ext/Language/en_us.lang.ext.php

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [ ] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: include/SugarObjects/LanguageManager.php [hunk 1]
```diff
@@ -202,6 +202,11 @@ public static function refreshLanguage($module, $lang, $loaded_mod_strings = arr
                     'custom/modules/'.$module.'/Ext/Language/'.$lang.'.lang.ext.php',
                  );
 
+        require_once('ModuleInstall/ModuleInstaller.php');
```
Comment (Dillon-Brown) [inline include/SugarObjects/LanguageManager.php]: Untrusted file inclusion:
```suggestion
        require_once __DIR__ . '/../../ModuleInstall/ModuleInstaller.php';
```

File: include/SugarObjects/LanguageManager.php [hunk 2]
```diff
@@ -202,6 +202,11 @@ public static function refreshLanguage($module, $lang, $loaded_mod_strings = arr
...
+        $mi = new ModuleInstaller();
+        $mi->modules = array($module);
```
Comment (Dillon-Brown) [inline include/SugarObjects/LanguageManager.php]: Short array syntax is preferred.
```suggestion
        $mi->modules = [($module)];
```

File: include/SugarObjects/LanguageManager.php [hunk 3]
```diff
@@ -202,6 +202,11 @@ public static function refreshLanguage($module, $lang, $loaded_mod_strings = arr
...
+        $mi->merge_files('Ext/Language/', $lang . '.lang.ext.php', $lang);
```
Comment (Dillon-Brown) [inline include/SugarObjects/LanguageManager.php]: Would this not have worrying security implications? I could pass in a local or remote PHP file and merge it with the ```lang.ext.php```.

File: include/utils/file_utils.php [hunk 4]
```diff
@@ -129,6 +129,27 @@ function write_array_to_file($the_name, $the_array, $the_file, $mode="w", $heade
     return sugar_file_put_contents($the_file, $the_string, LOCK_EX) !== false;
 }
 
+function write_override_label_to_file($the_name, $the_array, $the_file, $mode = "w", $header = '')
```
Comment (Dillon-Brown) [inline include/utils/file_utils.php]: Missing PHPDoc. Should also replace the double quotes with single.

File: include/utils/file_utils.php [hunk 5]
```diff
@@ -129,6 +129,27 @@ function write_array_to_file($the_name, $the_array, $the_file, $mode="w", $heade
...
+{
+    if (!empty($header) && ($mode != 'a' || !file_exists($the_file))) {
```
Comment (Dillon-Brown) [inline include/utils/file_utils.php]: Should use strict comparison.
```suggestion
    if (!empty($header) && ($mode !== 'a' || !file_exists($the_file))) {
```

File: include/utils/file_utils.php [hunk 6]
```diff
@@ -129,6 +129,27 @@ function write_array_to_file($the_name, $the_array, $the_file, $mode="w", $heade
...
+        $the_string = $header;
+    } else {
+        $the_string = "<?php\n" .
+            '// created: ' . date('Y-m-d H:i:s') . "\n";
+    }
+
+    foreach ($the_array as $labelName => $labelValue) {
+        $the_string .= "$" . "{$the_name}['{$labelName}'] = '{$labelValue}';\n";
```
Comment (Dillon-Brown) [inline include/utils/file_utils.php]:
```suggestion
        $the_string .= '$' . "{$the_name}['{$labelName}'] = '{$labelValue}';\n";
```

File: modules/ModuleBuilder/parsers/parser.label.php [hunk 7]
```diff
@@ -225,7 +219,7 @@ public static function addLabels($language, $labels, $moduleName, $basepath = nu
                 // obtain $mod_strings
                 include $filename;
             } elseif ($forRelationshipLabel) {
-                $fh = fopen($filename, 'ab');
+                $fh = fopen($filename, 'a');
```
Comment (Dillon-Brown) [inline modules/ModuleBuilder/parsers/parser.label.php]: Why was the binary-safe option removed here?

File: modules/ModuleBuilder/parsers/parser.label.php [hunk 8]
```diff
@@ -246,7 +240,7 @@ public static function addLabels($language, $labels, $moduleName, $basepath = nu
         if ($changed) {
             $GLOBALS [ 'log' ]->debug("ParserLabel::addLabels: writing new mod_strings to $filename");
             $GLOBALS [ 'log' ]->debug('ParserLabel::addLabels: mod_strings='.print_r($mod_strings, true));
-            if (!write_array_to_file('mod_strings', $mod_strings, $filename)) {
+            if (!write_override_label_to_file('mod_strings', $mod_strings, $filename)) {
```
Comment (Dillon-Brown) [inline modules/ModuleBuilder/parsers/parser.label.php]: The ```write_override_label_to_file``` is pretty much identical to ```write_array_to_file```. They should be combined so we don't have duplicated code.

File: include/SugarObjects/LanguageManager.php (hunk 3 above)
Comment (PedroErnst) [reply include/SugarObjects/LanguageManager.php]: Not exactly sure how this could be exploited.

Comment (Dillon-Brown) [reply include/SugarObjects/LanguageManager.php]: Debugged this further and couldn't find any obvious place that this could be injected. LGTM.

**Review by jack7anderson7 — DISMISSED** [review]

**Review by Dillon-Brown — APPROVED** [review]

**Review by jack7anderson7 — APPROVED** [review]

**Comment by jack7anderson7** [comment]
This fix was added in as part of #8476 this PR will be closed as a result.

**Pull request closed without merging**